bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java

```diff
@@ -152,6 +159,14 @@ public BookieRequestProcessor(ServerConfiguration serverCfg, Bookie bookie,
             createExecutor(
                 this.serverCfg.getNumLongPollWorkerThreads(),
                 "BookieLongPollThread-" + serverCfg.getBookiePort(), OrderedScheduler.NO_TASK_LIMIT);
+        this.fenceThreadPool = Executors.newCachedThreadPool(
```

Comment:
I don't think you need a separate thread pool if you fix #283. I would suggest fixing #283 first.

Reply:
#283 is to submit a runnable to the read processor thread for checking the fencing? I don't think this applies here. This fenceThreadPool is for V3. The problem was that writeThreadPool and readThreadPool can reject executions, if they hit a limit. This was causing failures in on of the tests for fencing (testRecoveryNotRejected). I would think that fencing should be outside of these execution limits, as it is part of recovery.

Reply:
ah sorry I read as it is for v2 process. I thought that you move fence read to a separate thread.

then I have a few questions here /cc @merlimat 

1) this change will move fence read to netty I/O thread, which it would be blocking subsequent requests. If read threads is accumulating requests, that typically means I/O is slow. fenceRead can potentially be very slow due to overwhelmed I/O, then it will subsequently block all the network I/O. I am not sure how much we can gain from moving fence read to netty I/O thread. It is against the purposes we having thread pools, is to avoid blocking network I/O. 

@merlimat any metrics or experiences show this change improved things a lot?

2) If this change is to prioritize recovery related ops, isn't that better to have its own recovery read thread pools? rather than blocking network I/O.

3) even worse at V2, the fence read is waiting for both read I/O and journal I/O (v3 makes fence request a non-blocking op on waiting fence result). If the bookie is overwhelmed (e.g. hitting bandwidth limit), journal can become a bottleneck and hence block netty I/O thread. (I assume this would potentially block sending responses back)

All the questions are related to the goal of not blocking network thread. If I read this change correctly, this change is against this goal, which can potentially have very bad impact.


then shall v2 fencing be executed in this fencing thread or not? because v2 fencing is actually a blocking call, while v3 fencing is not a blocking call. If you run v2 fencing in netty I/O thread, isn't it blocking other I/O operations?

Reply:
@sijie This change is a companion of #706. The story here is that in BK 4.6 if the read rate is very high, maxing out the bookie IO capacity what starts to happen is that read requests are piled up in the read executor task queue. 

The read thread is processing the read request, but since they spent a lot of time in the queue, the client will have already timed them out and re-sent other read requests. 

In #706, we've introduced the max-limit for the size of the executor queue, in order to limit the amount of time spent in the queue and rejecting additional requests in order to fail-fast. 

The rationale is to avoid to read ops that are going to be timed-out and that will not make any progress. 

Another problem with the read operations is that when the timeout/rejections happens, it impacts also fencing-reads which are used in recovery. 

This change is to make sure the recovery reads can make progress even though there is a very high rate of regular read requests. As Ivan pointed out, the recovery reads are critical because they typically prevent a writer from being operative, while regular reads are "just" reading data and they can thus be deferred to lower priority. 

The original change was doing the fence-read directly in IO thread as the way to skip the executor thread queue. I guess that using a dedicated thread pool for fence reads is probably a much better option in order not to block the IO threads.

Reply:
+1 on dedicated recovery thread pool. I'm currently working on BC stuff first though, but will come back once that's done.

Reply:
@merlimat +1 on a separate recovery thread pool.

Reply:
Separate thread pool does sound like a good idea, but I would really restrict recovery reads to real fencing reads only, not under replication reads.

Reply:
@sijie @jvrao @merlimat Trying to get back the context on this again. Any objection to reusing the fencing thread pool for the recovery reads/adds?